### PR TITLE
Refine the commit f9da39114d

### PR DESCRIPTION
**Pull request by AMD-dwang**
A better way to use Map.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4f0e99224e87e533385eef78f9cd55e2e0d6fca1
<details><summary>CTS tests (Failed: 0/138378)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35163/69163 (50.8%)</li>
<li>Failed: 0/69163 (0.0%)</li>
<li>Not Supported: 34000/69163 (49.2%)</li><li>Warnings: 0/69163 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35242/69215 (50.9%)</li>
<li>Failed: 0/69215 (0.0%)</li>
<li>Not Supported: 33973/69215 (49.1%)</li><li>Warnings: 0/69215 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/lower/llpcSpirvLowerConstImmediateStore.cpp [hunk 1]
```diff
@@ -223,16 +223,12 @@ bool SpirvLowerConstImmediateStore::tryProcessAlloca(AllocaInst *allocaInst) {
...
-  GlobalVariable *global = nullptr;
-  auto iter = m_allocToGlobals.find(initializer);
-  if (iter == m_allocToGlobals.end()) {
+  GlobalVariable *&global = m_allocToGlobals[initializer];
+  if (!global) {
     global = new GlobalVariable(*m_module, allocatedTy,
                                 true, // isConstant
                                 GlobalValue::InternalLinkage, initializer, "", nullptr, GlobalValue::NotThreadLocal,
                                 SPIRAS_Constant);
-    m_allocToGlobals[initializer] = global;
```
Comment (mariusz-sikora-at-amd) [inline llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:233]: Where is new global added to the map now ?

Comment (mariusz-sikora-at-amd) [reply llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:233]: You may run these lit tests (https://github.com/GPUOpen-Drivers/llpc/pull/2854) to verify the change.

Comment (jasilvanus) [reply llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:233]: The new code is taking a *reference* to the global pointer stored in the map, so the assignment `global = new...` updates the map.

**Review by jayfoad — APPROVED** [review]
Thanks!

File: llpc/lower/llpcSpirvLowerConstImmediateStore.cpp [hunk 2]
```diff
@@ -223,16 +223,12 @@ bool SpirvLowerConstImmediateStore::tryProcessAlloca(AllocaInst *allocaInst) {
   }
 
   // Step 3: Create the global variable and replace the alloca
-  GlobalVariable *global = nullptr;
-  auto iter = m_allocToGlobals.find(initializer);
-  if (iter == m_allocToGlobals.end()) {
+  GlobalVariable *&global = m_allocToGlobals[initializer];
+  if (!global) {
```
Comment (Flakebi) [inline llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:227]: This assumes that if the key `initializer` was not found in `m_allocToGlobals`, a new entry is created and the value of the new entry is `nullptr`, right?

`DensedMap::InsertIntoBucket` creates the new value with `::new (&TheBucket->getSecond()) ValueT(std::forward<ValueArgs>(Values)...);` (where `Values` is empty), so it calls the default constructor of `GlobalVariable*`.
If I am not mistaken, the default constructor of a pointer **does not initialize memory**, so checking for `nullptr` afterwards only works when we are lucky.

`try_emplace(initializer, nullptr)` should do the correct thing.

Comment (jayfoad) [reply llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:227]: > But as DenseMap doesn't explicitly document any of this, using it conservatively won't hurt.

My feeling is that DenseMap behaves exactly like std::map in this regard, so it's fine to rely on the value initialization.

Comment (Flakebi) [reply llpc/lower/llpcSpirvLowerConstImmediateStore.cpp:227]: That value initialization is a good point. This code looks like it does the correct thing then :)

**Comment by tsymalla-AMD** [comment]
Thanks, but I'd rather edit the commit message (title + message) at least, since it is not really informative without looking up the explicit commit.

**Review by Flakebi — APPROVED** [review]

**Comment by Flakebi** [comment]
You can edit the commit title and message when you ‘squash and merge’.

**Pull request merged**